### PR TITLE
k8s - fix issue with server side apply

### LINKED ISSUE #548: Server side apply broken in kubernetes >= 25.3.0
##### SUMMARY

Something changed in version 25 of the kubernetes library. I don't see anything in the changelog, but using server side apply now fails when the kubernetes library version is 25.3.0+. The likely candidate is https://github.com/ansible-collections/kubernetes.core/blob/69304d1c3b1d9cce8aba9343c4da433191be4ce9/plugins/module_utils/apply.py#L134.

See: https://github.com/ansible-collections/kubernetes.core/issues/547#issuecomment-1332554709

##### ISSUE TYPE
- Bug Report

##### COLLECTION VERSION

```paste below
2.3.2
```

##### STEPS TO REPRODUCE

The following playbook works as expected when kubernetes==24.2.0 and fails when kubernetes=25.3.0.

```yaml
- hosts: localhost
  gather_facts: false
  tasks:
    - kubernetes.core.k8s:
        apply: true
        server_side_apply:
          field_manager: ansible
        definition:
          kind: Namespace
          apiVersion: v1
          metadata:
            name: testing

```

##### ACTUAL RESULTS

```paste below
The full traceback is:
Traceback (most recent call last):
  File "/home/mgraves/.ansible/tmp/ansible-tmp-1669834036.478591-2018817-17689802905430/AnsiballZ_k8s.py", line 107, in <module>
    _ansiballz_main()
  File "/home/mgraves/.ansible/tmp/ansible-tmp-1669834036.478591-2018817-17689802905430/AnsiballZ_k8s.py", line 99, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/mgraves/.ansible/tmp/ansible-tmp-1669834036.478591-2018817-17689802905430/AnsiballZ_k8s.py", line 47, in invoke_module
    runpy.run_module(mod_name='ansible_collections.kubernetes.core.plugins.modules.k8s', init_globals=dict(_module_fqn='ansible_collections.kubernetes.core.plugins.modules.k8s', _modlib_path=modlib_path),
  File "/usr/lib/python3.9/runpy.py", line 210, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_kubernetes.core.k8s_payload_ppk312hc/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s.py", line 486, in <module>
  File "/tmp/ansible_kubernetes.core.k8s_payload_ppk312hc/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s.py", line 482, in main
  File "/tmp/ansible_kubernetes.core.k8s_payload_ppk312hc/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s.py", line 459, in execute_module
  File "/tmp/ansible_kubernetes.core.k8s_payload_ppk312hc/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/common.py", line 807, in execute_module
  File "/tmp/ansible_kubernetes.core.k8s_payload_ppk312hc/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/common.py", line 1114, in perform_action
  File "/tmp/ansible_kubernetes.core.k8s_payload_ppk312hc/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/k8sdynamicclient.py", line 45, in apply
  File "/tmp/ansible_kubernetes.core.k8s_payload_ppk312hc/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/apply.py", line 136, in k8s_apply
  File "/home/mgraves/git/ansible/venv-3.9/lib/python3.9/site-packages/kubernetes/dynamic/client.py", line 164, in server_side_apply
    return self.request('patch', path, body=body, force_conflicts=force_conflicts, **kwargs)
  File "/home/mgraves/git/ansible/venv-3.9/lib/python3.9/site-packages/kubernetes/dynamic/client.py", line 55, in inner
    resp = func(self, *args, **kwargs)
  File "/home/mgraves/git/ansible/venv-3.9/lib/python3.9/site-packages/kubernetes/dynamic/client.py", line 270, in request
    api_response = self.client.call_api(
  File "/home/mgraves/git/ansible/venv-3.9/lib/python3.9/site-packages/kubernetes/client/api_client.py", line 348, in call_api
    return self.__call_api(resource_path, method,
  File "/home/mgraves/git/ansible/venv-3.9/lib/python3.9/site-packages/kubernetes/client/api_client.py", line 180, in __call_api
    response_data = self.request(
  File "/home/mgraves/git/ansible/venv-3.9/lib/python3.9/site-packages/kubernetes/client/api_client.py", line 407, in request
    return self.rest_client.PATCH(url,
  File "/home/mgraves/git/ansible/venv-3.9/lib/python3.9/site-packages/kubernetes/client/rest.py", line 296, in PATCH
    return self.request("PATCH", url,
  File "/home/mgraves/git/ansible/venv-3.9/lib/python3.9/site-packages/kubernetes/client/rest.py", line 168, in request
    request_body = json.dumps(body)
  File "/usr/lib/python3.9/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.9/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.9/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type bytes is not JSON serializable
```

### PR DESCRIPTION
**Pull request by abikouo**
##### SUMMARY

Fix #548 and #547

##### ISSUE TYPE

- Bugfix Pull Request

### CODE REVIEW

**Comment by abikouo** [comment]
recheck

File: plugins/module_utils/apply.py [hunk 1]
```diff
@@ -131,14 +138,18 @@ def k8s_apply(resource, definition, **kwargs):
     existing, desired = apply_object(resource, definition)
     server_side = kwargs.get("server_side", False)
     if server_side:
-        body = json.dumps(definition).encode()
+        versions = gather_versions()
+        body = definition
+        if LooseVersion(versions["kubernetes"]) < LooseVersion("25.0.3"):
```
Comment (gravesm) [inline plugins/module_utils/apply.py:143]: I'm assuming you meant to put 25.3.0 here? We should probably just use 25.0.0 though, as this is likely more correct.

Comment (abikouo) [reply plugins/module_utils/apply.py:143]: Yes, will do

File: plugins/module_utils/apply.py [hunk 2]
```diff
@@ -131,14 +138,18 @@ def k8s_apply(resource, definition, **kwargs):
...
+            body = json.dumps(definition).encode()
         # server_side_apply is forces content_type to 'application/apply-patch+yaml'
         return resource.server_side_apply(
             body=body,
             name=definition["metadata"]["name"],
             namespace=definition["metadata"].get("namespace"),
             force_conflicts=kwargs.get("force_conflicts"),
             field_manager=kwargs.get("field_manager"),
+            dry_run=kwargs.get("dry_run"),
```
[the remark this answers is not in the document]
Comment (abikouo) [reply plugins/module_utils/apply.py:152]: done

File: changelogs/fragments/549-fix-server-side-apply.yaml [hunk 3]
```diff
@@ -1,3 +1,4 @@
 ---
 bugfixes:
   - k8s - Fix issue with server side apply with kubernetes release '25.3.0' (https://github.com/ansible-collections/kubernetes.core/issues/548).
+  - k8s - Fix issue with server side apply with kubernetes release '25.3.0' (https://github.com/ansible-collections/kubernetes.core/issues/547).
```
Comment (gravesm) [inline changelogs/fragments/549-fix-server-side-apply.yaml:4]: Could you change the description to address that this fixes check mode? This problem does not have anything to do with the kubernetes version.

Comment (abikouo) [reply changelogs/fragments/549-fix-server-side-apply.yaml]: Copy/paste issue. Fixed
Thanks

**Review by gravesm — APPROVED** [review]

**Comment by loganmc10** [comment]
This is a blocking issue for me, is it possible to release a new version with this fix included?

**Comment by stiller-leser** [comment]
Same here, this is blocking for us. Any ETA on a new release?

**Pull request merged**